java-squid/src/main/java/org/sonar/java/se/AbstractBinaryConstraint.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2016 SonarSource SA
+ * mailto:contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.se;
+
+import org.sonar.java.se.ConstraintManager.BooleanConstraint;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+public abstract class AbstractBinaryConstraint {
+
+  public static AbstractBinaryConstraint contraintBetween(SymbolicValue v1, SymbolicValue v2, Collection<AbstractBinaryConstraint> constraints) {
+    for (AbstractBinaryConstraint constraint : constraints) {
+      AbstractBinaryConstraint c = constraint.getConstraintBetween(v1, v2, constraints);
+      if (c != null) {
+        return c;
+      }
+    }
+    return null;
+  }
+
+  protected final SymbolicValue v1;
+  protected final SymbolicValue v2;
+
+  protected AbstractBinaryConstraint(SymbolicValue v1, SymbolicValue v2) {
+    this.v1 = v1;
+    this.v2 = v2;
+  }
+
+  @Override
+  public boolean equals(Object obj) {
+    if (getClass().equals(obj.getClass())) {
```

Comment:
Moreover you should probably rely on some helpers to properly implements equals and hashcode. See https://github.com/SonarSource/sonar-java/blob/master/java-squid/src/main/java/org/sonar/java/se/ProgramState.java#L147 for instance.